javalib/src/main/scala/java/util/regex/PatternCompiler.scala

```diff
@@ -0,0 +1,1645 @@
+              }
+              pIndex += 2 // for the \E
+              -1
+
+            case _ =>
+              parseSingleCodePointEscape()
+          }
+
+        case _ =>
+          pIndex += charCount(dispatchCP)
+          dispatchCP
+      }
+
+      if (startCodePoint != -1) {
+        @inline def canBeRangeEnd(c: Char): Boolean = c != '[' && c != ']'
+
+        if (pIndex + 2 <= len && pattern.charAt(pIndex) == '-' &&
+            canBeRangeEnd(pattern.charAt(pIndex + 1))) {
+          // Range of code points
+
+          pIndex += 1
+          if (pIndex + 2 > len)
+            parseError("Unclosed character class")
+
+          val cpEnd = pattern.codePointAt(pIndex)
+          pIndex += charCount(cpEnd)
+          val endCodePoint = if (cpEnd == '\\') {
+            parseSingleCodePointEscape()
+          } else {
+            cpEnd
+          }
+
+          if (endCodePoint < startCodePoint)
+            parseError("Illegal character range")
+
+          addCodePointRange(startCodePoint, endCodePoint)
+        } else {
+          // Single code point
+          addSingleCodePoint(startCodePoint)
+        }
+      }
+    }
+
+    parseError("Unclosed character class")
+    // scalastyle:on return
+  }
+
+  private def compileGroup(): String = {
+    val pattern = this.pattern // local copy
+    val len = pattern.length()
+
+    val start = pIndex
+
+    if (start + 1 == len || pattern.charAt(start + 1) != '?') {
+      pIndex = start + 1
+      originalGroupCount += 1
+      compiledGroupCount += 1
+      groupNumberMap.push(compiledGroupCount)
+      "(" + compileInsideGroup() + ")"
+    } else {
+      if (start + 2 == len)
+        parseError("Unclosed group")
+
+      val c1 = pattern.charAt(start + 2)
+
+      if (c1 == ':' || c1 == '=' || c1 == '!') {
+        pIndex = start + 3
+        pattern.substring(start, start + 3) + compileInsideGroup() + ")"
+      } else if (c1 == '<') {
+        if (start + 3 == len)
+          parseError("Unclosed group")
+
+        val c2 = pattern.charAt(start + 3)
+
+        if (isLetter(c2)) {
+          val nameStart = start + 3
+          var nameEnd = nameStart + 1
+          while (nameEnd != len && isLetterOrDigit(pattern.charAt(nameEnd)))
+            nameEnd += 1
+          if (nameEnd == len || pattern.charAt(nameEnd) != '>')
+            parseError("named capturing group is missing trailing '>'")
+          val name = pattern.substring(nameStart, nameEnd)
+          pIndex = nameEnd + 1
+          originalGroupCount += 1
+          compiledGroupCount += 1
+          groupNumberMap.push(compiledGroupCount)
+          if (namedGroups.contains(name))
+            parseError(s"named capturing group <$name> is already defined")
+          namedGroups(name) = originalGroupCount
+          "(" + compileInsideGroup() + ")"
+        } else {
+          requireES2018Features("Look-behind group")
+          if (c2 != '=' && c2 != '!')
+            parseError("Unknown look-behind group")
```

Comment:
I'd put this error before the feature error. (basically: say I don't know what you want rather then, guess and say I can't give you what you want and once you could, say, actually, I don't know what you want).